main/src/main/scala/sbt/internal/CrossJava.scala

```diff
@@ -0,0 +1,365 @@
+                "configuration. This could result in subprojects cross building against Java versions that they are " +
+                "not compatible with. Try issuing cross building command with tasks instead, since sbt will be able " +
+                "to ensure that cross building is only done using configured project and Java version combinations " +
+                "that are configured."
+            )
+            state.log.debug("Java versions configuration is:")
+            projCrossVersions.foreach {
+              case (project, versions) => state.log.debug(s"$project: $versions")
+            }
+          }
+
+          // Execute using a blanket switch
+          projCrossVersions.toMap.apply(currentRef).flatMap { version =>
+            // Force scala version
+            Seq(s"$JavaSwitchCommand $verbose $version!", aggCommand)
+          }
+
+        case Right(_) =>
+          // We have a key, we're likely to be able to cross build this using the per project behaviour.
+
+          // Group all the projects by scala version
+          projVersions.groupBy(_._2).mapValues(_.map(_._1)).toSeq.flatMap {
+            case (version, Seq(project)) =>
+              // If only one project for a version, issue it directly
+              Seq(s"$JavaSwitchCommand $verbose $version", s"$project/$aggCommand")
+            case (version, projects) if aggCommand.contains(" ") =>
+              // If the command contains a space, then the `all` command won't work because it doesn't support issuing
+              // commands with spaces, so revert to running the command on each project one at a time
+              s"$JavaSwitchCommand $verbose $version" :: projects
+                .map(project => s"$project/$aggCommand")
+            case (version, projects) =>
+              // First switch scala version, then use the all command to run the command on each project concurrently
+              Seq(
+                s"$JavaSwitchCommand $verbose $version",
+                projects.map(_ + "/" + aggCommand).mkString("all ", " ", "")
+              )
+          }
+      }
+
+      allCommands.toList ::: captureCurrentSession(state, x)
+    }
+  }
+
+  private val JavaCapturedSession = AttributeKey[Seq[Setting[_]]]("javaCrossCapturedSession")
+
+  private def captureCurrentSession(state: State, extracted: Extracted): State = {
+    state.put(JavaCapturedSession, extracted.session.rawAppend)
+  }
+
+  def discoverJavaHomes: ListMap[String, File] = {
+    import JavaDiscoverConfig._
+    val configs = Vector(jabba, linux, macOS)
+    ListMap(configs flatMap { _.javaHomes }: _*)
+  }
+
+  sealed trait JavaDiscoverConf {
+    def javaHomes: Vector[(String, File)]
+  }
+
+  object JavaDiscoverConfig {
+    val linux = new JavaDiscoverConf {
+      val base: File = file("/usr") / "lib" / "jvm"
+      val JavaHomeDir = """java-([0-9]+)-.*""".r
```

Comment:
FYI this regex won't match on a default Fedora Open JDK install. See https://github.com/akka/akka/issues/25298